safe/gui/tools/minimum_needs/needs_calculator_dialog.py

```diff
@@ -61,148 +98,113 @@ def __init__(self, parent=None):
-        layer = QgsMapLayerRegistry.instance().mapLayer(layer_id)
-        fields = layer.pendingFields()
-        self.cboFields.clear()
-        has_fields = False
-        for field in fields:
-            # LOGGER.info(field.typeName())
-            # TODO exclude dates too? TS
-            if field.typeName() != 'String':
-                has_fields = True
-                add_ordered_combo_item(
-                    self.cboFields, field.name(), field.name())
-
-        # Now disable the run button if no suitable fields were found
-        # see #2206
-        ok_button = self.button_box.button(QtGui.QDialogButtonBox.Ok)
-        if not has_fields:
-            ok_button.setEnabled(False)
-        else:
-            ok_button.setEnabled(True)
+        # create memory layer
+        output_layer_name = os.path.splitext(input_layer.name())[0]
+        output_layer = (
+            '%s_minimum_needs' % output_layer_name)
+        output_layer = create_memory_layer(
+            output_layer,
+            input_layer.geometryType(),
+            input_layer.crs(),
+            input_layer.fields())
+
+        # monkey patching input layer to make it work with
+        # prepare vector layer function
+        temp_layer = input_layer
+        temp_layer.keywords = {
+            'layer_purpose': layer_purpose_aggregation['key']}
+
+        # copy features to output layer
+        copy_layer(temp_layer, output_layer)
+
+        # Monkey patching output layer to make it work with
+        # minimum needs calculator
+        output_layer.keywords['layer_purpose'] = (
+            layer_purpose_aggregation['key'])
+        output_layer.keywords['inasafe_fields'] = (
+            {displaced_field['key']: self.displaced.currentField(),
+             aggregation_name_field['key']:
+                 self.aggregation_name.currentField()})
+
+        # remove unnecessary fields & rename inasafe fields
+        rename_remove_inasafe_fields(output_layer)
+
+        return output_layer
 
     def accept(self):
         """Process the layer and field and generate a new layer.
 
         .. note:: This is called on OK click.
 
         """
-        index = self.cboFields.currentIndex()
-        field_name = self.cboFields.itemData(
-            index, QtCore.Qt.UserRole)
-
-        index = self.cboPolygonLayers.currentIndex()
-        layer_id = self.cboPolygonLayers.itemData(
-            index, QtCore.Qt.UserRole)
-        # noinspection PyArgumentList
-        layer = QgsMapLayerRegistry.instance().mapLayer(layer_id)
-
-        file_name = layer.source()
 
         try:
-            output_layer = self.minimum_needs(input_layer, field_name)
-        except ValueError:
+            self.result_layer = self.minimum_needs(self.layer.currentLayer())
+        except Exception as e:
+            LOGGER.debug(e)
+            display_critical_message_box(
+                title='Error while calculating minimum needs', message=e)
```

Comment:
But there is another issue.
`e` is a python exception object.
You are trying to give the exception as a message but the function expects a string : https://github.com/inasafe/inasafe/blob/develop/safe/utilities/qgis_utilities.py#L259

I just tried this in the QGIS python console : 
```
from safe.utilities.qgis_utilities import display_critical_message_box
import logging

LOGGER = logging.getLogger('InaSAFE')

try:
    raise TypeError
except Exception as e:
    LOGGER.exception(e)
    display_critical_message_box(title='Error while calculating minimum needs', message=e)
```
* The logger didn't log the exception correctly. I though it was going to be ok. Maybe you should just log an error.
* This code will generate another exception because you tried to to give an exception instead of a string. Let's only display that some thing went wrong.
```
display_critical_message_box(title='Error while calculating minimum needs', message=e)
  File "/Users/etienne/.qgis2/python/plugins/inasafe/safe/utilities/qgis_utilities.py", line 262, in display_critical_message_box
    QMessageBox.critical(parent, title, message)
TypeError: arguments did not match any overloaded call:
  QMessageBox.critical(QWidget, QString, QString, QMessageBox.StandardButtons buttons=QMessageBox.Ok, QMessageBox.StandardButton defaultButton=QMessageBox.NoButton): argument 3 has unexpected type 'exceptions.TypeError'
  QMessageBox.critical(QWidget, QString, QString, int, int, int button2=0): argument 3 has unexpected type 'exceptions.TypeError'
  QMessageBox.critical(QWidget, QString, QString, QString, QString button1Text=QString(), QString button2Text=QString(), int defaultButtonNumber=0, int escapeButtonNumber=-1): argument 3 has unexpected type 'exceptions.TypeError'
```
You can use `e.__class__.__name__` to get the name of the exception as a string. It might be enough. @ismailsunni ?

I just remember that the postprocessor function will return a boolean with a error message.
https://github.com/inasafe/inasafe/blob/develop/safe/impact_function/postprocessors.py#L62
I think you should check if the boolean if True. If not true, you can display this error message.

@ismailsunni Any ideas where to get the error message ?